test/index.test.ts

```diff
@@ -14,6 +36,81 @@ interface MockUser {
   email: string;
 }
 
+type ReturnTypeFromStage<
+  T extends (context: any) => any,
+  TStage extends AllStageKeys
+> = TStage extends AllAsyncStageKeys
+  ? PromiseResolveOrSync<ReturnType<T>>
+  : ReturnType<T>;
+
+async function HTPipeTest<
+  TPipe extends OptionallyHasInitPreContext<any, any> &
+    OptionallyHasSanitizeParams<any, any> &
+    OptionallyHasSanitizeBody<any, any> &
+    MightHavePreAuthorize<any, any> &
+    OptionallyHasAttachData<any, any> &
+    MightHaveFinalAuthorize<any, any> &
+    OptionallyHasDoWork<any, any> &
+    MightHaveRespond<any, any> &
+    MightHaveSanitizeResponse<any, any>,
```

Comment:
This isn't quite what we discussed in chat...
- [ ] why is TPipeIn extends TPipeInExpected?
- [ ] why is TLifecycleStage dependent on what TPipe HAPPENS to extend in some order?  Shouldn't it just be based on TPipe and TStage?

Reply:
TPipeIn extends TPipeInExpected because of pipeInExpected param
now since i removed that param we don't need that anymore
other one done

Reply:
just in case
TPipe extends HasInitPreContext<any, any>
      ? TPipe[TStage]
      : never
it was in () but prettier remove them